File: client-android/lib/src/main/kotlin/stasis/client_android/lib/model/server/devices/DeviceBootstrapParameters.kt
package stasis.client_android.lib.model.server.devices

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class DeviceBootstrapParameters(
    val authentication: Authentication,
    @field:Json(name = "server_api")
    val serverApi: ServerApi,
    @field:Json(name = "server_core")
    val serverCore: ServerCore,
    val secrets: SecretsConfig
) {
    @JsonClass(generateAdapter = true)
    data class Authentication(
        @field:Json(name = "token_endpoint")
        val tokenEndpoint: String,
        @field:Json(name = "client_id")
        val clientId: String,
        @field:Json(name = "client_secret")
        val clientSecret: String,
        val scopes: Scopes
    )

    @JsonClass(generateAdapter = true)
    data class ServerApi(
        val url: String,
        val user: String,
        @field:Json(name = "user_salt")
        val userSalt: String,
        val device: String
    )

    @JsonClass(generateAdapter = true)
    data class ServerCore(
        val address: String,
        @field:Json(name = "node_id")
        val nodeId: String
    )

    @JsonClass(generateAdapter = true)
    data class Scopes(
        val api: String,
        val core: String
    )

    @JsonClass(generateAdapter = true)
    data class SecretsConfig(
        val derivation: Derivation,
        val encryption: Encryption
    ) {
        @JsonClass(generateAdapter = true)
        data class Derivation(
            val encryption: Derivation.Encryption,
            val authentication: Derivation.Authentication
        ) {
            @JsonClass(generateAdapter = true)
            data class Encryption(
                @field:Json(name = "secret_size")
                val secretSize: Int,
                val iterations: Int,
                @field:Json(name = "salt_prefix")
                val saltPrefix: String
            )

            @JsonClass(generateAdapter = true)
            data class Authentication(
                val enabled: Boolean,
                @field:Json(name = "secret_size")
                val secretSize: Int,
                val iterations: Int,
                @field:Json(name = "salt_prefix")
                val saltPrefix: String
            )
        }

        @JsonClass(generateAdapter = true)
        data class Encryption(
            val file: Encryption.File,
            val metadata: Encryption.Metadata,
            @field:Json(name = "device_secret")
            val deviceSecret: Encryption.DeviceSecret
        ) {
            @JsonClass(generateAdapter = true)
            data class File(
                @field:Json(name = "key_size")
                val keySize: Int,
                @field:Json(name = "iv_size")
                val ivSize: Int
            )

            @JsonClass(generateAdapter = true)
            data class Metadata(
                @field:Json(name = "key_size")
                val keySize: Int,
                @field:Json(name = "iv_size")
                val ivSize: Int
            )

            @JsonClass(generateAdapter = true)
            data class DeviceSecret(
                @field:Json(name = "key_size")
                val keySize: Int,
                @field:Json(name = "iv_size")
                val ivSize: Int
            )
        }
    }
}

File: client-ios/lib/Sources/StasisClientLib/Model/Server/Devices/DeviceBootstrapParameters.swift
import Foundation

public struct DeviceBootstrapParameters: Sendable, Equatable, Hashable, Codable {
    public let authentication: Authentication
    public let serverApi: ServerApi
    public let serverCore: ServerCore
    public let secrets: SecretsConfig

    public init(
        authentication: Authentication,
        serverApi: ServerApi,
        serverCore: ServerCore,
        secrets: SecretsConfig
    ) {
        self.authentication = authentication
        self.serverApi = serverApi
        self.serverCore = serverCore
        self.secrets = secrets
    }

    public struct Authentication: Sendable, Equatable, Hashable, Codable {
        public let tokenEndpoint: String
        public let clientId: String
        public let clientSecret: String
        public let scopes: Scopes

        public init(tokenEndpoint: String, clientId: String, clientSecret: String, scopes: Scopes) {
            self.tokenEndpoint = tokenEndpoint
            self.clientId = clientId
            self.clientSecret = clientSecret
            self.scopes = scopes
        }
    }

    public struct ServerApi: Sendable, Equatable, Hashable, Codable {
        public let url: String
        public let user: String
        public let userSalt: String
        public let device: String

        public init(url: String, user: String, userSalt: String, device: String) {
            self.url = url
            self.user = user
            self.userSalt = userSalt
            self.device = device
        }
    }

    public struct ServerCore: Sendable, Equatable, Hashable, Codable {
        public let address: String
        public let nodeId: String

        public init(address: String, nodeId: String) {
            self.address = address
            self.nodeId = nodeId
        }
    }

    public struct Scopes: Sendable, Equatable, Hashable, Codable {
        public let api: String
        public let core: String

        public init(api: String, core: String) {
            self.api = api
            self.core = core
        }
    }

    public struct SecretsConfig: Sendable, Equatable, Hashable, Codable {
        public let derivation: Derivation
        public let encryption: Encryption

        public init(derivation: Derivation, encryption: Encryption) {
            self.derivation = derivation
            self.encryption = encryption
        }

        public struct Derivation: Sendable, Equatable, Hashable, Codable {
            public let encryption: Encryption
            public let authentication: Authentication

            public init(encryption: Encryption, authentication: Authentication) {
                self.encryption = encryption
                self.authentication = authentication
            }

            public struct Encryption: Sendable, Equatable, Hashable, Codable {
                public let secretSize: Int
                public let iterations: Int
                public let saltPrefix: String

                public init(secretSize: Int, iterations: Int, saltPrefix: String) {
                    self.secretSize = secretSize
                    self.iterations = iterations
                    self.saltPrefix = saltPrefix
                }
            }

            public struct Authentication: Sendable, Equatable, Hashable, Codable {
                public let enabled: Bool
                public let secretSize: Int
                public let iterations: Int
                public let saltPrefix: String

                public init(enabled: Bool, secretSize: Int, iterations: Int, saltPrefix: String) {
                    self.enabled = enabled
                    self.secretSize = secretSize
                    self.iterations = iterations
                    self.saltPrefix = saltPrefix
                }
            }
        }

        public struct Encryption: Sendable, Equatable, Hashable, Codable {
            public let file: File
            public let metadata: Metadata
            public let deviceSecret: DeviceSecret

            public init(file: File, metadata: Metadata, deviceSecret: DeviceSecret) {
                self.file = file
                self.metadata = metadata
                self.deviceSecret = deviceSecret
            }

            public struct File: Sendable, Equatable, Hashable, Codable {
                public let keySize: Int
                public let ivSize: Int

                public init(keySize: Int, ivSize: Int) {
                    self.keySize = keySize
                    self.ivSize = ivSize
                }
            }

            public struct Metadata: Sendable, Equatable, Hashable, Codable {
                public let keySize: Int
                public let ivSize: Int

                public init(keySize: Int, ivSize: Int) {
                    self.keySize = keySize
                    self.ivSize = ivSize
                }
            }

            public struct DeviceSecret: Sendable, Equatable, Hashable, Codable {
                public let keySize: Int
                public let ivSize: Int

                public init(keySize: Int, ivSize: Int) {
                    self.keySize = keySize
                    self.ivSize = ivSize
                }
            }
        }
    }
}
